File: src/node/generator/lib/spanner/spanner.js
const http = require('http');

class SpannerGenerator {
    constructor() {
        this.sessionName = null;
    }

    async init() {
        for (let i = 0; i < 30; i++) {
            try {
                await this.makeRequest('http://spanner:9020/', 'GET');
                break;
            } catch (err) {
                console.log("Waiting for Spanner emulator...");
                await new Promise(resolve => setTimeout(resolve, 2000));
            }
        }

        try {
            const dbName = "projects/test-project/instances/test-instance/databases/test-db";
            const response = await this.makeRequest(`http://spanner:9020/v1/${dbName}/sessions`, 'POST', {});
            const data = JSON.parse(response);
            if (data.name) {
                this.sessionName = data.name;
                console.log(`Successfully created Spanner session: ${this.sessionName}`);
            } else {
                console.error("Session name not found in response:", data);
                process.exit(1);
            }
        } catch (err) {
            console.error("Failed to initialize Spanner session:", err);
            process.exit(1);
        }
    }

    makeRequest(url, method, body = null) {
        return new Promise((resolve, reject) => {
            const { hostname, port, pathname } = new URL(url);
            const options = {
                hostname,
                port: port || 80,
                path: pathname,
                method,
                headers: {
                    'Content-Type': 'application/json'
                }
            };

            const req = http.request(options, (res) => {
                let data = '';
                res.on('data', (chunk) => {
                    data += chunk;
                });
                res.on('end', () => {
                    if (res.statusCode >= 200 && res.statusCode < 300) {
                        resolve(data);
                    } else if (res.statusCode === 404 && method === 'GET') {
                        resolve(data); // Emulator returns 404 for root GET but it means it's up
                    } else {
                        reject(new Error(`Request failed with status code ${res.statusCode}: ${data}`));
                    }
                });
            });

            req.on('error', (err) => {
                reject(err);
            });

            if (body) {
                req.write(JSON.stringify(body));
            }
            req.end();
        });
    }

    async nextIdString() {
        if (!this.sessionName) {
            console.error("Spanner session not initialized.");
            return "";
        }

        try {
            const url = `http://spanner:9020/v1/${this.sessionName}:executeSql`;
            const body = { sql: "SELECT GET_NEXT_SEQUENCE_VALUE(SEQUENCE global_id_seq)" };
            const response = await this.makeRequest(url, 'POST', body);
            const data = JSON.parse(response);

            if (data.rows && data.rows.length > 0 && data.rows[0].length > 0) {
                return data.rows[0][0];
            }
            console.error("Failed to extract sequence value from response:", data);
            return "";
        } catch (err) {
            console.error("Error executing SQL:", err);
            return "";
        }
    }
}

module.exports = SpannerGenerator;

File: src/python/generator/lib/spanner/spanner.py
import urllib.request
import urllib.error
import json
import time
import sys

class SpannerGenerator:
    def __init__(self):
        self.session_name = None

        for i in range(30):
            try:
                req = urllib.request.Request('http://spanner:9020/', method='GET')
                with urllib.request.urlopen(req) as response:
                    if response.status in (200, 404):
                        break
            except urllib.error.HTTPError as e:
                if e.code == 404:
                    break
            except Exception as e:
                print("Waiting for Spanner emulator...", flush=True)
                time.sleep(2)

        try:
            db_name = "projects/test-project/instances/test-instance/databases/test-db"
            url = f"http://spanner:9020/v1/{db_name}/sessions"
            data = json.dumps({}).encode('utf-8')
            req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
            
            with urllib.request.urlopen(req) as response:
                res_data = json.loads(response.read().decode('utf-8'))
                if 'name' in res_data:
                    self.session_name = res_data['name']
                    print(f"Successfully created Spanner session: {self.session_name}", flush=True)
                else:
                    print(f"Session name not found in response: {res_data}", file=sys.stderr)
                    sys.exit(1)
        except Exception as e:
            print(f"Failed to initialize Spanner session: {e}", file=sys.stderr)
            sys.exit(1)

    def next_id_string(self):
        if not self.session_name:
            print("Spanner session not initialized.", file=sys.stderr)
            return ""

        try:
            url = f"http://spanner:9020/v1/{self.session_name}:executeSql"
            data = json.dumps({"sql": "SELECT GET_NEXT_SEQUENCE_VALUE(SEQUENCE global_id_seq)"}).encode('utf-8')
            req = urllib.request.Request(url, data=data, headers={'Content-Type': 'application/json'}, method='POST')
            
            with urllib.request.urlopen(req) as response:
                res_data = json.loads(response.read().decode('utf-8'))
                if 'rows' in res_data and len(res_data['rows']) > 0 and len(res_data['rows'][0]) > 0:
                    return str(res_data['rows'][0][0])
            print(f"Failed to extract sequence value from response: {res_data}", file=sys.stderr)
            return ""
        except Exception as e:
            print(f"Error executing SQL: {e}", file=sys.stderr)
            return ""
